src/main/java/net/glowstone/entity/GlowEntity.java

```diff
@@ -957,28 +969,50 @@ protected void pulsePhysics() {
             if (velLocZ.getBlock().getType().isOccluding()) {
                 velocity.setZ(0);
             }
+            collide(block);
         } else {
-            // apply friction and gravity
-            if (location.getBlock().getType() == Material.WATER) {
-                velocity.multiply(liquidDrag);
-                velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
-            } else if (location.getBlock().getType() == Material.LAVA) {
-                velocity.multiply(liquidDrag - 0.3);
-                velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
-            } else {
-                velocity.setY(airDrag * (velocity.getY() + getGravityAccel().getY()));
-                if (isOnGround()) {
-                    velocity.setX(velocity.getX() * slipMultiplier);
-                    velocity.setZ(velocity.getZ() * slipMultiplier);
+            if (hasFriction()) {
+                // apply friction and gravity
+                if (location.getBlock().getType() == Material.WATER) {
+                    velocity.multiply(liquidDrag);
+                    velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
+                } else if (location.getBlock().getType() == Material.LAVA) {
+                    velocity.multiply(liquidDrag - 0.3);
+                    velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
                 } else {
-                    velocity.setX(velocity.getX() * 0.91);
-                    velocity.setZ(velocity.getZ() * 0.91);
+                    velocity.setY(airDrag * (velocity.getY() + getGravityAccel().getY()));
+                    if (isOnGround()) {
+                        velocity.setX(velocity.getX() * slipMultiplier);
+                        velocity.setZ(velocity.getZ() * slipMultiplier);
+                    } else {
+                        horizontalAirDrag = 0.91;
+                        velocity.setX(velocity.getX() * horizontalAirDrag);
+                        velocity.setZ(velocity.getZ() * 0.91);
+                    }
+                }
+            } else if (hasGravity() && !isOnGround()) {
+                switch (location.getBlock().getType()) {
+                    case WATER:
+                    case LAVA:
+                        velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
+                        break;
+                    default:
+                        velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
```

Comment:
Gravity should not be divided by 4 here.

Reply:
That's the existing behavior (see existing lines 964 and 967).